lib/galaxy/tool_util/deps/container_resolvers/explicit.py

```diff
@@ -54,6 +56,64 @@ def resolve(self, enabled_container_types, tool_info, **kwds):
         return None
 
 
+class CachedExplicitSingularityContainerResolver(CliContainerResolver):
+    resolver_type = 'cached_explicit_singularity'
+    container_type = 'singularity'
+    cli = 'singularity'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.cache_directory_path = kwargs.get("cache_directory", os.path.join(kwargs['app_info'].container_image_cache_path, "singularity", "explicit"))
+        self._init_cache_directory()
+
+    def _init_cache_directory(self):
+        os.makedirs(self.cache_directory_path, exist_ok=True)
+
+    def resolve(self, enabled_container_types, tool_info, **kwds):
+        """Find a container explicitly mentioned in tool description.
+
+        This ignores the tool requirements and assumes the tool author crafted
+        a correct container. We use singularity here to fetch docker containers,
+        hence the container_description hack here.
+        """
+        for container_description in tool_info.container_descriptions:  # type: ContainerDescription
+            if container_description.type == 'docker':
```

Comment:
Wondering if we want ro support other URIs like `shub://vsoch/singularity-images`?

Reply:
That works already. This code is taken from the normal explicit resolver.
If `container_description.type == 'singularity'` any valid singularity container will work (such as `shub://vsoch/singularity-images`).